File: tests/julia/test_data_validation.jl
"""
Layer 1: Data File Validation Tests

Pure schema and sanity checks on the Vietnam JSON data files.
Runs WITHOUT a solver — only reads and validates the data layer.

Run: julia --project tests/julia/test_data_validation.jl
"""

using Test
using JSON

const REPO_ROOT = abspath(joinpath(@__DIR__, "..", ".."))
const DATA_DIR = joinpath(REPO_ROOT, "data", "vietnam")
const MANIFEST_PATH = joinpath(DATA_DIR, "manifest.json")

# ---------------------------------------------------------------------------
# Helper: load a data file by manifest key
# ---------------------------------------------------------------------------
function load_data_file(manifest::Dict, key::String)
    filename = manifest[key]
    filepath = joinpath(DATA_DIR, filename)
    return JSON.parsefile(filepath), filename
end

# ---------------------------------------------------------------------------
# Load manifest once
# ---------------------------------------------------------------------------
const MANIFEST = JSON.parsefile(MANIFEST_PATH)

@testset "Layer 1: Vietnam Data File Validation" begin

    # ===================================================================
    # 1. Manifest structure
    # ===================================================================
    @testset "Manifest structure" begin
        required_keys = ["tariff", "tech_costs", "financials", "emissions", "export_rules"]
        for k in required_keys
            @test haskey(MANIFEST, k)
        end

        for k in required_keys
            filename = MANIFEST[k]
            filepath = joinpath(DATA_DIR, filename)
            @test isfile(filepath)
        end
    end

    # ===================================================================
    # 2. Schema compliance — every file has _meta + data
    # ===================================================================
    @testset "Schema compliance (_meta + data)" begin
        required_meta_fields = ["version", "effective_date", "source", "last_updated"]

        for key in ["tariff", "tech_costs", "financials", "emissions", "export_rules"]
            raw, filename = load_data_file(MANIFEST, key)

            @testset "$filename" begin
                @test haskey(raw, "_meta")
                @test haskey(raw, "data")

                if haskey(raw, "_meta")
                    meta = raw["_meta"]
                    for field in required_meta_fields
                        @test haskey(meta, field)
                    end
                end
            end
        end
    end

    # ===================================================================
    # 3. Tariff sanity
    # ===================================================================
    @testset "Tariff sanity" begin
        raw, _ = load_data_file(MANIFEST, "tariff")
        data = raw["data"]

        @testset "Base price" begin
            @test haskey(data, "base_avg_price_vnd_per_kwh")
            base = data["base_avg_price_vnd_per_kwh"]
            @test base > 0
            @test base < 100_000
        end

        @testset "TOU schedule completeness" begin
            schedule = data["tou_schedule"]
            @test haskey(schedule, "weekday")
            sunday_key = haskey(schedule, "sunday") ? "sunday" : "sunday_and_public_holidays"
            @test haskey(schedule, sunday_key)

            for (day_type, block) in [("weekday", schedule["weekday"]), (sunday_key, schedule[sunday_key])]
                all_hours = Int[]
                for period in ["peak_hours", "standard_hours", "offpeak_hours"]
                    @test haskey(block, period)
                    append!(all_hours, Int.(block[period]))
                end
                sorted = sort(unique(all_hours))
                @test sorted == collect(0:23)
            end
        end

        @testset "Rate multipliers — industrial" begin
            mults = data["rate_multipliers"]
            @test haskey(mults, "industrial")

            for (vl_name, vl) in mults["industrial"]
                vl isa Dict || continue
                @testset "$vl_name" begin
                    @test haskey(vl, "peak")
                    @test haskey(vl, "standard")
                    @test haskey(vl, "offpeak")
                    @test vl["peak"] > vl["standard"] > vl["offpeak"]
                    @test vl["peak"] > 0
                    @test vl["offpeak"] > 0
                end
            end
        end

        @testset "Rate multipliers — commercial" begin
            mults = data["rate_multipliers"]
            @test haskey(mults, "commercial")

            for (subcategory, subcat_rates) in mults["commercial"]
                subcat_rates isa Dict || continue
                for (vl_name, vl) in subcat_rates
                    vl isa Dict || continue
                    @testset "$subcategory/$vl_name" begin
                        @test haskey(vl, "peak")
                        @test haskey(vl, "standard")
                        @test haskey(vl, "offpeak")
                        @test vl["peak"] > vl["standard"] > vl["offpeak"]
                    end
                end
            end
        end

        @testset "Rate multipliers — household tiers" begin
            mults = data["rate_multipliers"]
            @test haskey(mults, "household")
            hh = mults["household"]
            tier_keys = filter(k -> startswith(k, "tier_"), collect(keys(hh)))
            @test length(tier_keys) >= 2
            for tk in tier_keys
                @test hh[tk] > 0
                @test hh[tk] < 5.0
            end
        end
    end

    # ===================================================================
    # 4. Tech cost bounds
    # ===================================================================
    @testset "Tech cost bounds" begin
        raw, _ = load_data_file(MANIFEST, "tech_costs")
        data = raw["data"]

        @testset "PV costs" begin
            @test haskey(data, "PV")
            for pv_type in ["rooftop", "ground"]
                if haskey(data["PV"], pv_type)
                    for region in ["north", "central", "south"]
                        if haskey(data["PV"][pv_type], region)
                            cost = data["PV"][pv_type][region]["installed_cost_per_kw"]
                            @test 200 <= cost <= 2000
                            om = data["PV"][pv_type][region]["om_cost_per_kw"]
                            @test om >= 0
                        end
                    end
                end
            end
        end

        @testset "Wind costs" begin
            if haskey(data, "Wind")
                for wind_type in ["onshore"]
                    if haskey(data["Wind"], wind_type)
                        for region in ["north", "central", "south"]
                            if haskey(data["Wind"][wind_type], region)
                                cost = data["Wind"][wind_type][region]["installed_cost_per_kw"]
                                @test 500 <= cost <= 5000
                            end
                        end
                    end
                end
            end
        end

        @testset "Battery costs" begin
            if haskey(data, "ElectricStorage")
                es = data["ElectricStorage"]
                if haskey(es, "li_ion")
                    for region in ["north", "central", "south"]
                        if haskey(es["li_ion"], region)
                            cost_kw = es["li_ion"][region]["installed_cost_per_kw"]
                            cost_kwh = es["li_ion"][region]["installed_cost_per_kwh"]
                            @test cost_kw > 0
                            @test 50 <= cost_kwh <= 1000
                        end
                    end
                end
                if haskey(es, "common_defaults")
                    @test es["common_defaults"]["installed_cost_constant"] == 0
                end
            end
        end

        @testset "PV common defaults — zero incentives" begin
            if haskey(data["PV"], "common_defaults")
                cd = data["PV"]["common_defaults"]
                @test cd["federal_itc_fraction"] == 0
                @test cd["macrs_option_years"] == 0
                @test cd["macrs_bonus_fraction"] == 0
                @test cd["state_ibi_fraction"] == 0
                @test cd["utility_ibi_fraction"] == 0
            end
        end
    end

    # ===================================================================
    # 5. Emissions factor range
    # ===================================================================
    @testset "Emissions factor" begin
        raw, _ = load_data_file(MANIFEST, "emissions")
        data = raw["data"]

        @test haskey(data, "grid_emission_factor_lb_CO2_per_kwh")
        ef = data["grid_emission_factor_lb_CO2_per_kwh"]
        @test 0.0 < ef <= 3.0

        @test haskey(data, "grid_emission_factor_tCO2e_per_mwh")
        ef_t = data["grid_emission_factor_tCO2e_per_mwh"]
        @test 0.0 < ef_t <= 2.0

        # Cross-check: lb/kWh ≈ tCO2e/MWh × 2204.62 / 1000
        expected_lb = ef_t * 2204.62 / 1000
        @test abs(ef - expected_lb) < 0.01

        @test haskey(data, "series_type")
        if data["series_type"] == "constant"
            @test haskey(data, "series_length")
            @test data["series_length"] == 8760
        end
    end

    # ===================================================================
    # 6. Financial bounds
    # ===================================================================
    @testset "Financial bounds" begin
        raw, _ = load_data_file(MANIFEST, "financials")
        data = raw["data"]

        for profile_name in ["standard", "renewable_energy_preferential", "high_tech_zone"]
            if haskey(data, profile_name)
                @testset "$profile_name" begin
                    p = data[profile_name]

                    if haskey(p, "offtaker_tax_rate_fraction")
                        @test 0 <= p["offtaker_tax_rate_fraction"] <= 1
                    end
                    if haskey(p, "owner_tax_rate_fraction")
                        @test 0 <= p["owner_tax_rate_fraction"] <= 1
                    end
                    if haskey(p, "offtaker_discount_rate_fraction")
                        @test 0 < p["offtaker_discount_rate_fraction"] <= 1
                    end
                    if haskey(p, "owner_discount_rate_fraction")
                        @test 0 < p["owner_discount_rate_fraction"] <= 1
                    end
                    if haskey(p, "elec_cost_escalation_rate_fraction")
                        @test -0.1 <= p["elec_cost_escalation_rate_fraction"] <= 0.2
                    end
                    if haskey(p, "om_cost_escalation_rate_fraction")
                        @test -0.1 <= p["om_cost_escalation_rate_fraction"] <= 0.2
                    end
                    if haskey(p, "analysis_years")
                        @test 1 <= p["analysis_years"] <= 50
                    end
                end
            end
        end
    end

    # ===================================================================
    # 7. Export rules
    # ===================================================================
    @testset "Export rules" begin
        raw, _ = load_data_file(MANIFEST, "export_rules")
        data = raw["data"]

        @testset "Rooftop solar" begin
            @test haskey(data, "rooftop_solar")
            rs = data["rooftop_solar"]
            @test haskey(rs, "max_export_fraction")
            @test 0 < rs["max_export_fraction"] <= 1
            @test haskey(rs, "surplus_purchase_rate_vnd_per_kwh")
            @test rs["surplus_purchase_rate_vnd_per_kwh"] > 0
            @test haskey(rs, "surplus_purchase_rate_usd_per_kwh")
            @test rs["surplus_purchase_rate_usd_per_kwh"] > 0
        end

        @testset "REopt mapping" begin
            @test haskey(data, "reopt_mapping")
            rm = data["reopt_mapping"]
            @test rm["can_net_meter"] == false
            @test rm["can_wholesale"] == true
            @test rm["can_export_beyond_nem_limit"] == false
        end

        @testset "DPPA ceiling tariffs" begin
            if haskey(data, "dppa_ceiling_tariffs_vnd_per_kwh")
                dppa = data["dppa_ceiling_tariffs_vnd_per_kwh"]
                for (tech, regions) in dppa
                    tech == "notes" && continue
                    regions isa Dict || continue
                    for (region, rate) in regions
                        @test rate > 0
                    end
                end
            end
        end
    end

end # top-level testset

println("\n✓ Layer 1: All data validation tests completed.")

File: tests/python/reopt/test_data_validation.py
"""
Layer 1: Data File Validation Tests (Python)

Pure schema and sanity checks on the Vietnam JSON data files.
Runs WITHOUT a solver — only reads and validates the data layer.
Mirror of tests/julia/test_data_validation.jl — identical checks, same data files.

Run: pytest tests/python/reopt/test_data_validation.py -v
"""

import json
from pathlib import Path

import pytest

REPO_ROOT = Path(__file__).resolve().parent.parent.parent.parent
DATA_DIR = REPO_ROOT / "data" / "vietnam"
MANIFEST_PATH = DATA_DIR / "manifest.json"

# ---------------------------------------------------------------------------
# Fixtures
# ---------------------------------------------------------------------------


@pytest.fixture(scope="module")
def manifest():
    with open(MANIFEST_PATH, "r", encoding="utf-8") as f:
        return json.load(f)


def _load_data_file(manifest: dict, key: str):
    filename = manifest[key]
    filepath = DATA_DIR / filename
    with open(filepath, "r", encoding="utf-8") as f:
        return json.load(f), filename


# ===================================================================
# 1. Manifest structure
# ===================================================================


class TestManifestStructure:
    REQUIRED_KEYS = ["tariff", "tech_costs", "financials", "emissions", "export_rules"]

    def test_required_keys_present(self, manifest):
        for k in self.REQUIRED_KEYS:
            assert k in manifest, f"manifest.json missing required key: {k}"

    def test_referenced_files_exist(self, manifest):
        for k in self.REQUIRED_KEYS:
            filename = manifest[k]
            filepath = DATA_DIR / filename
            assert filepath.is_file(), f"File not found: {filepath}"


# ===================================================================
# 2. Schema compliance — every file has _meta + data
# ===================================================================


class TestSchemaCompliance:
    REQUIRED_META_FIELDS = ["version", "effective_date", "source", "last_updated"]

    @pytest.mark.parametrize(
        "key", ["tariff", "tech_costs", "financials", "emissions", "export_rules"]
    )
    def test_meta_and_data_blocks(self, manifest, key):
        raw, filename = _load_data_file(manifest, key)
        assert "_meta" in raw, f"{filename} missing _meta block"
        assert "data" in raw, f"{filename} missing data block"

        meta = raw["_meta"]
        for field in self.REQUIRED_META_FIELDS:
            assert field in meta, f"{filename}._meta missing field: {field}"


# ===================================================================
# 3. Tariff sanity
# ===================================================================


class TestTariffSanity:
    @pytest.fixture(scope="class")
    def tariff_data(self, manifest):
        raw, _ = _load_data_file(manifest, "tariff")
        return raw["data"]

    def test_base_price_positive(self, tariff_data):
        assert "base_avg_price_vnd_per_kwh" in tariff_data
        base = tariff_data["base_avg_price_vnd_per_kwh"]
        assert base > 0
        assert base < 100_000

    @pytest.mark.parametrize("day_type", ["weekday", "sunday_and_public_holidays"])
    def test_tou_schedule_completeness(self, tariff_data, day_type):
        schedule = tariff_data["tou_schedule"]
        assert day_type in schedule
        block = schedule[day_type]

        all_hours = []
        for period in ["peak_hours", "standard_hours", "offpeak_hours"]:
            assert period in block, f"{day_type} missing {period}"
            all_hours.extend(int(h) for h in block[period])

        assert sorted(set(all_hours)) == list(range(24)), (
            f"{day_type} TOU hours don't cover 0-23: got {sorted(set(all_hours))}"
        )

    def test_industrial_multipliers(self, tariff_data):
        mults = tariff_data["rate_multipliers"]
        assert "industrial" in mults

        for vl_name, vl in mults["industrial"].items():
            if not isinstance(vl, dict):
                continue
            assert "peak" in vl, f"industrial.{vl_name} missing peak"
            assert "standard" in vl, f"industrial.{vl_name} missing standard"
            assert "offpeak" in vl, f"industrial.{vl_name} missing offpeak"
            assert vl["peak"] > vl["standard"] > vl["offpeak"], (
                f"industrial.{vl_name}: peak > standard > offpeak violated"
            )
            assert vl["peak"] > 0
            assert vl["offpeak"] > 0

    def test_commercial_multipliers(self, tariff_data):
        mults = tariff_data["rate_multipliers"]
        assert "commercial" in mults

        for subcategory, subcat_rates in mults["commercial"].items():
            if not isinstance(subcat_rates, dict):
                continue
            for vl_name, vl in subcat_rates.items():
                if not isinstance(vl, dict):
                    continue
                assert "peak" in vl, f"commercial.{subcategory}.{vl_name} missing peak"
                assert "standard" in vl, (
                    f"commercial.{subcategory}.{vl_name} missing standard"
                )
                assert "offpeak" in vl, (
                    f"commercial.{subcategory}.{vl_name} missing offpeak"
                )
                assert vl["peak"] > vl["standard"] > vl["offpeak"]

    def test_household_tiers(self, tariff_data):
        mults = tariff_data["rate_multipliers"]
        assert "household" in mults
        hh = mults["household"]
        tier_keys = [k for k in hh if k.startswith("tier_")]
        assert len(tier_keys) >= 2
        for tk in tier_keys:
            assert hh[tk] > 0
            assert hh[tk] < 5.0


# ===================================================================
# 4. Tech cost bounds
# ===================================================================


class TestTechCostBounds:
    @pytest.fixture(scope="class")
    def tech_data(self, manifest):
        raw, _ = _load_data_file(manifest, "tech_costs")
        return raw["data"]

    def test_pv_costs(self, tech_data):
        assert "PV" in tech_data
        for pv_type in ["rooftop", "ground"]:
            if pv_type not in tech_data["PV"]:
                continue
            for region in ["north", "central", "south"]:
                if region not in tech_data["PV"][pv_type]:
                    continue
                cost = tech_data["PV"][pv_type][region]["installed_cost_per_kw"]
                assert 200 <= cost <= 2000, (
                    f"PV {pv_type}/{region} cost {cost} out of range"
                )
                om = tech_data["PV"][pv_type][region]["om_cost_per_kw"]
                assert om >= 0

    def test_wind_costs(self, tech_data):
        if "Wind" not in tech_data:
            pytest.skip("No Wind data")
        for wind_type in ["onshore"]:
            if wind_type not in tech_data["Wind"]:
                continue
            for region in ["north", "central", "south"]:
                if region not in tech_data["Wind"][wind_type]:
                    continue
                cost = tech_data["Wind"][wind_type][region]["installed_cost_per_kw"]
                assert 500 <= cost <= 5000, (
                    f"Wind {wind_type}/{region} cost {cost} out of range"
                )

    def test_battery_costs(self, tech_data):
        if "ElectricStorage" not in tech_data:
            pytest.skip("No ElectricStorage data")
        es = tech_data["ElectricStorage"]
        if "li_ion" in es:
            for region in ["north", "central", "south"]:
                if region not in es["li_ion"]:
                    continue
                cost_kw = es["li_ion"][region]["installed_cost_per_kw"]
                cost_kwh = es["li_ion"][region]["installed_cost_per_kwh"]
                assert cost_kw > 0
                assert 50 <= cost_kwh <= 1000, (
                    f"Battery {region} cost_kwh {cost_kwh} out of range"
                )
        if "common_defaults" in es:
            assert es["common_defaults"]["installed_cost_constant"] == 0

    def test_pv_common_defaults_zero_incentives(self, tech_data):
        if "common_defaults" not in tech_data["PV"]:
            pytest.skip("No PV common_defaults")
        cd = tech_data["PV"]["common_defaults"]
        assert cd["federal_itc_fraction"] == 0
        assert cd["macrs_option_years"] == 0
        assert cd["macrs_bonus_fraction"] == 0
        assert cd["state_ibi_fraction"] == 0
        assert cd["utility_ibi_fraction"] == 0


# ===================================================================
# 5. Emissions factor range
# ===================================================================


class TestEmissionsFactor:
    @pytest.fixture(scope="class")
    def emissions_data(self, manifest):
        raw, _ = _load_data_file(manifest, "emissions")
        return raw["data"]

    def test_lb_co2_per_kwh_range(self, emissions_data):
        assert "grid_emission_factor_lb_CO2_per_kwh" in emissions_data
        ef = emissions_data["grid_emission_factor_lb_CO2_per_kwh"]
        assert 0.0 < ef <= 3.0

    def test_tco2e_per_mwh_range(self, emissions_data):
        assert "grid_emission_factor_tCO2e_per_mwh" in emissions_data
        ef_t = emissions_data["grid_emission_factor_tCO2e_per_mwh"]
        assert 0.0 < ef_t <= 2.0

    def test_unit_cross_check(self, emissions_data):
        ef = emissions_data["grid_emission_factor_lb_CO2_per_kwh"]
        ef_t = emissions_data["grid_emission_factor_tCO2e_per_mwh"]
        expected_lb = ef_t * 2204.62 / 1000
        assert abs(ef - expected_lb) < 0.01

    def test_series_type(self, emissions_data):
        assert "series_type" in emissions_data
        if emissions_data["series_type"] == "constant":
            assert "series_length" in emissions_data
            assert emissions_data["series_length"] == 8760


# ===================================================================
# 6. Financial bounds
# ===================================================================


class TestFinancialBounds:
    @pytest.fixture(scope="class")
    def financial_data(self, manifest):
        raw, _ = _load_data_file(manifest, "financials")
        return raw["data"]

    @pytest.mark.parametrize(
        "profile_name", ["standard", "renewable_energy_preferential", "high_tech_zone"]
    )
    def test_profile_bounds(self, financial_data, profile_name):
        if profile_name not in financial_data:
            pytest.skip(f"Profile {profile_name} not in data")
        p = financial_data[profile_name]

        if "offtaker_tax_rate_fraction" in p:
            assert 0 <= p["offtaker_tax_rate_fraction"] <= 1
        if "owner_tax_rate_fraction" in p:
            assert 0 <= p["owner_tax_rate_fraction"] <= 1
        if "offtaker_discount_rate_fraction" in p:
            assert 0 < p["offtaker_discount_rate_fraction"] <= 1
        if "owner_discount_rate_fraction" in p:
            assert 0 < p["owner_discount_rate_fraction"] <= 1
        if "elec_cost_escalation_rate_fraction" in p:
            assert -0.1 <= p["elec_cost_escalation_rate_fraction"] <= 0.2
        if "om_cost_escalation_rate_fraction" in p:
            assert -0.1 <= p["om_cost_escalation_rate_fraction"] <= 0.2
        if "analysis_years" in p:
            assert 1 <= p["analysis_years"] <= 50


# ===================================================================
# 7. Export rules
# ===================================================================


class TestExportRules:
    @pytest.fixture(scope="class")
    def export_data(self, manifest):
        raw, _ = _load_data_file(manifest, "export_rules")
        return raw["data"]

    def test_rooftop_solar(self, export_data):
        assert "rooftop_solar" in export_data
        rs = export_data["rooftop_solar"]
        assert "max_export_fraction" in rs
        assert 0 < rs["max_export_fraction"] <= 1
        assert "surplus_purchase_rate_vnd_per_kwh" in rs
        assert rs["surplus_purchase_rate_vnd_per_kwh"] > 0
        assert "surplus_purchase_rate_usd_per_kwh" in rs
        assert rs["surplus_purchase_rate_usd_per_kwh"] > 0

    def test_reopt_mapping(self, export_data):
        assert "reopt_mapping" in export_data
        rm = export_data["reopt_mapping"]
        assert rm["can_net_meter"] is False
        assert rm["can_wholesale"] is True
        assert rm["can_export_beyond_nem_limit"] is False

    def test_dppa_ceiling_tariffs(self, export_data):
        if "dppa_ceiling_tariffs_vnd_per_kwh" not in export_data:
            pytest.skip("No DPPA ceiling tariffs")
        dppa = export_data["dppa_ceiling_tariffs_vnd_per_kwh"]
        for tech, regions in dppa.items():
            if tech == "notes" or not isinstance(regions, dict):
                continue
            for region, rate in regions.items():
                assert rate > 0, f"DPPA {tech}/{region} rate must be > 0"
